src/map.cpp

```diff
@@ -943,68 +943,36 @@ void QTreeLeafNode::removeCreature(Creature* c)
 uint32_t Map::clean() const
 {
 	uint64_t start = OTSYS_TIME();
-	size_t count = 0, tiles = 0;
+	size_t tiles = 0;
 
 	if (g_game.getGameState() == GAME_STATE_NORMAL) {
 		g_game.setGameState(GAME_STATE_MAINTAIN);
 	}
 
-	std::vector<const QTreeNode*> nodes {
-		&root
-	};
 	std::vector<Item*> toRemove;
-	do {
-		const QTreeNode* node = nodes.back();
-		nodes.pop_back();
-		if (node->isLeaf()) {
-			const QTreeLeafNode* leafNode = static_cast<const QTreeLeafNode*>(node);
-			for (uint8_t z = 0; z < MAP_MAX_LAYERS; ++z) {
-				Floor* floor = leafNode->getFloor(z);
-				if (!floor) {
-					continue;
-				}
-
-				for (auto& row : floor->tiles) {
-					for (auto tile : row) {
-						if (!tile || tile->hasFlag(TILESTATE_PROTECTIONZONE)) {
-							continue;
-						}
 
-						TileItemVector* itemList = tile->getItemList();
-						if (!itemList) {
-							continue;
-						}
-
-						++tiles;
-						for (Item* item : *itemList) {
-							if (item->isCleanable()) {
-								toRemove.push_back(item);
-							}
-						}
-
-						for (Item* item : toRemove) {
-							g_game.internalRemoveItem(item, -1);
-						}
-						count += toRemove.size();
-						toRemove.clear();
-					}
-				}
-			}
-		} else {
-			for (auto childNode : node->child) {
-				if (childNode) {
-					nodes.push_back(childNode);
-				}
+	for (auto tileList : g_game.getCleanTiles()) {
+		++tiles;
+		for (Item* item : *tileList->getItemList()) {
+			if (item->isCleanable()) {
+				toRemove.emplace_back(item);
 			}
 		}
-	} while (!nodes.empty());
+	}
+
+	for (auto item : toRemove) {
+		g_game.internalRemoveItem(item, -1);
+	}
 
 	if (g_game.getGameState() == GAME_STATE_MAINTAIN) {
 		g_game.setGameState(GAME_STATE_NORMAL);
 	}
 
+	size_t count = toRemove.size();
+	g_game.clearCleanTiles();
```

Comment:
This really should happen before the game state is changed to normal.

Reply:
moved it up, thanks